accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1303,12 +1296,7 @@ rather than from `Object`)*.
 
 *This change is needed because some extension types are subtypes of
 `Object?` and not subtypes of `Object`, and they need to have a
-well-defined depth. Note that the depth of an extension type can be
-determined by its actual type arguments, if any, because type parameters of
-the extension type may occur in its representation type. In particular, the
-depth of an extension type is a property of the type itself, and it is not
-always possible to determine the depth from the associated extension type
-declaration.*
+well-defined depth.*
```

Comment:
But this means that the depth of an extension type is defined if it has `Object` as an explicitly declared superinterface (directly or indirectly, via other extension types or via other non-extension types), and it is undefined otherwise.

Reply:
In which case we can define *depth* for extension types with no non-extension type superinterfaces too, without needing to change changing the super-interface hierarchy.

> Let `T` be a type declared by of a class, mixin, enum or extension type declaration. let *S* be the set of direct > superinterfaces of `T`. Then we define *depth*(`T`) as:
> * *depth*(`T`) = 0 if *S* is empty. `T` has no direct superinterfaces. _This includes `Object`, `Null` and any extension type without an `implements` clause._
> * *depth*(`T`) = max{`I` in *S*| *depth*(`I`)} + 1 if *S* is not empty.

This defines a *depth* function on all the nominative types, without introducing `Object?` as an interface (which it isn't) and without touching the superinterface relation. Minimal change, but enough.
The only other needed change is that LUB can now fail to find a shared superinterface.
And that's OK, in that case **UP** is `Object?`.

Still, *all* changes are entirely local to the LUB algorithm (which I believe is the only thing using *depth*), so there should be no risk of accidental or undesired side-effects elsewhere.

Reply:
We can define a notion of depth, but it doesn't stop there. We need a special exception to handle the choice of a result: It will no longer work to say "take the singleton set with the greatest depth", we need to say "take the singleton set with the greatest depth; if no such set exists then take `Object?`". For example, if we have an extension type with no `<interfaces>` and some other type then the entire set of superinterfaces is empty; we still follow the rule and take `Object?`, but the explanation involving empty sets and a special exceptional treatment of `Object?` is simply not intuitive.